# Data Cleaning with Pandas

In [1]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

In [2]:
!cd data

In [3]:
!ls

README.md                    data
Real Property Sales.zip      pandas-3-data-cleaning.ipynb
Real Property Sales.zip.1    wget-log
Residential Building.zip     wget-log.1


In [4]:
#!brew install wget

In [5]:
#!wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip

In [6]:
#!wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
    #web client that runs on command line that can download files

In [7]:
!cd data

In [8]:
!ls

README.md                    data
Real Property Sales.zip      pandas-3-data-cleaning.ipynb
Real Property Sales.zip.1    wget-log
Residential Building.zip     wget-log.1


### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [9]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [10]:
sales_df.T.head()
#0 for SalePrice: sometimes properties transfer/get inherited without a financial transaction
#for various circumstances, which we'd want to ignore

0           1           2           3           4        \
ExciseTaxNbr     2687551     1235111     2704079     2584094     1056831   
Major             138860      664885      423943      403700      951120   
Minor                110          40          50         715         900   
DocumentDate  08/21/2014  07/09/1991  10/11/2014  01/04/2013  04/20/1989   
SalePrice         245000           0           0           0       85000   

                 5           6           7           8           9        \
ExciseTaxNbr     1971467     1813750     1813750     1813750     1813750   
Major             605475       22209       22209       22209       22209   
Minor                 10        9001        9003        9007        9002   
DocumentDate  07/07/2003  04/25/2001  04/25/2001  04/25/2001  04/25/2001   
SalePrice              0     5755000     5755000     5755000     5755000   

                 ...         2014326     2014327     2014328     2014329  \
ExciseTaxNbr     ...         2184061     2184061     2066401     2066401   
Major            ...          340470      340470       46470      276770   
Minor            ...              31          32         790         505   
DocumentDate     ...      01/26/2006  01/26/2006  08/16/2004  08/16/2004   
SalePrice        ...          868000      868000      206650      206650   

                 2014330     2014331     2014332     2014333     2014334  \
ExciseTaxNbr     2616041     2213184     2899631     2200523     1541705   
Major             524980      401990      397170      375890      393190   
Minor               4024         175         835         225         455   
DocumentDate  06/13/2013  05/30/2006  11/02/2017  04/13/2006  04/24/1997   
SalePrice              0           0      445000           0      305000   

                 2014335  
ExciseTaxNbr     2492216  
Major             111300  
Minor                250  
DocumentDate  05/18/2011  
SalePrice              0  

[5 rows x 2014336 columns]

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [11]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [12]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [13]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [14]:
bldg_df.T

0                            1       \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                         2                  

### Another warning! Which column has index 11?

In [15]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [16]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [17]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [18]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [19]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [20]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [21]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [22]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [23]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [24]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [25]:
sales_df.info()
#float64 more usable int64 when there are NaNs

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [26]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [27]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [28]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [31]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [32]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]
#Boolean mask: squiggle flips 1's and 0's
#equation above is saying:
#keep sales_data dataframe where ROWS = ZipCode is NOT NaN, ALL COLUMNS

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [33]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1322148 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1322148 non-null float64
Minor            1322148 non-null float64
DocumentDate     1322148 non-null object
SalePrice        1322148 non-null int64
SqFtTotLiving    1322148 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 70.6+ MB


# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [34]:
sales_data = sales_data.loc[sales_data['SalePrice'] > 0]

In [35]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
6  423943.0   50.0   07/15/1999      96000            960   98092
7  423943.0   50.0   01/08/2001     127500            960   98092

In [36]:
sales_data = sales_data.loc[sales_data['SqFtTotLiving'] >= 100]

In [37]:
sales_data['SqFtTotLiving'].min()

100

In [38]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867397 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867397 non-null float64
Minor            867397 non-null float64
DocumentDate     867397 non-null object
SalePrice        867397 non-null int64
SqFtTotLiving    867397 non-null int64
ZipCode          867397 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [39]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode']

10168      98042-3001
10169      98042-3001
10170      98042-3001
10171      98042-3001
16528      98199-3014
16529      98199-3014
16530      98199-3014
40183      98031-3173
43081      98042-3001
43082      98042-3001
53691      98033-4917
53692      98033-4917
53693      98033-4917
53694      98033-4917
53695      98033-4917
87152      98058-9018
87153      98058-9018
93627      98136-1728
93628      98136-1728
93629      98136-1728
93630      98136-1728
146181     98058-7983
146183     98058-7983
151057     98074-6315
151058     98074-6315
151062     98074-6315
163641     98052-1963
242179     98023-2712
242181     98023-2712
251279     98028-4768
              ...    
1190128    98042-3001
1198332    98074-4092
1252275    98042-3001
1252667    98146-3008
1252668    98146-3008
1259514    98059-7120
1268080    98042-3001
1268081    98042-3001
1300491            WA
1300492            WA
1310974    98042-8215
1310975    98042-8215
1310976    98042-8215
1310977    98042-8215
1317353   

In [ ]:
integer_zip_df = sales_data.loc[sales_data['ZipCode'].apply(is_integer) == True]
not_integer_zip_df = sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False]

In [67]:
not_integer_zip_df['ZipCode'].apply(lambda x: x[:5]).values

array(['98042', '98042', '98042', '98042', '98199', '98199', '98199',
       '98031', '98042', '98042', '98033', '98033', '98033', '98033',
       '98033', '98058', '98058', '98136', '98136', '98136', '98136',
       '98058', '98058', '98074', '98074', '98074', '98052', '98023',
       '98023', '98028', '98028', '98028', '98028', '98038', '98038',
       '98038', '98177', '98177', 'A', 'A', '98042', '98042', '98074',
       '98074', '98074', '98074', '98074', '98074', '98074', '98059',
       '98034', '98034', 'B', ' ', ' ', ' ', '98023', '98023', '98028',
       '98028', '98074', '98074', '98006', '98053', '98074', '98074',
       '98008', '98008', '98075', '98075', '98075', '98075', '98075',
       '98055', '98055', '98042', '98042', '98075', '98118', '98118',
       '98118', '98074', '98074', ' ', ' ', '98028', '98198', '98198',
       '98198', '98198', '98075', '98075', '98075', '98075', '98028',
       '98028', '98028', '98107', '98107', '98107', '98074', '98074',
       '98028', 

In [47]:
sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'] = not_integer_zip_df['ZipCode'].apply(lambda x: x[:5]).values


In [60]:
sales_data = sales_data.loc[sales_data['ZipCode'].apply(is_integer) == True]


In [75]:
sales_data = sales_data.reset_index(drop=True)
sales_data

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0       138860.0   110.0   08/21/2014     245000           1490   98002   
1       138860.0   110.0   06/12/1989     109300           1490   98002   
2       138860.0   110.0   01/16/2005      14684           1490   98002   
3       423943.0    50.0   07/15/1999      96000            960   98092   
4       423943.0    50.0   01/08/2001     127500            960   98092   
5       403700.0   715.0   07/03/2013     464500           1780   98008   
6       403700.0   715.0   02/21/2013     357000           1780   98008   
7       403700.0   715.0   10/13/1995     142000           1780   98008   
8       403700.0   715.0   02/22/2007     528000           1780   98008   
9        98400.0   380.0   03/23/1994     139950           2050   98058   
10       98400.0   380.0   03/29/2017     439500           2050   98058   
11      330395.0  1610.0   06/07/2012     335000           2210   98038   
12      330395.0  1610.0   08/26/2013     369900           2210   98038   
13      330395.0  1610.0   06/27/1994     192950           2210   98038   
14      330395.0  1610.0   08/22/2013     369900           2210   98038   
15      330395.0  1610.0   11/10/1993     107500           2210   98038   
16      675260.0   940.0   03/06/1998     209950           2040   98031   
17      675260.0   940.0   08/01/1995     191950           2040   98031   
18      675260.0   940.0   09/13/2004     289950           2040   98031   
19      795508.0   170.0   04/09/2001     174000            980   98058   
20      795508.0   170.0   10/31/1996     135000            980   98058   
21      795508.0   170.0   04/30/2004     210000            980   98058   
22      795508.0   170.0   12/21/2012     148000            980   98058   
23      638580.0   110.0   07/14/2015     190000           2130   98188   
24      638580.0   110.0   12/21/2015     338000           2130   98188   
25      252206.0  9011.0   02/21/1997     300000           2530   98051   
26      252206.0  9011.0   11/28/2000     390000           2530   98051   
27      321165.0    50.0   05/10/2004    2100000           1800   98001   
28      321165.0    50.0   10/24/2013     215000           1800   98001   
29      321165.0    50.0   09/20/2004     270000           1800   98001   
...          ...     ...          ...        ...            ...     ...   
867356  928170.0     5.0   12/11/1986      89950           1630   98116   
867357  783280.0    60.0   08/21/1986      67500           1200   98146   
867358  809260.0    80.0   08/20/2001     145000           1170   98042   
867359  344800.0  1500.0   07/12/1996     174000            680   98125   
867360  177421.0   370.0   06/24/1993     233000           2930   98077   
867361  278731.0   900.0   10/18/1982      78000           1590   98031   
867362   12605.0  9008.0   10/01/1993     110000            790   98072   
867363   42506.0  9158.0   03/07/1988      35000           1920   98053   
867364  955020.0  4120.0   03/24/1994     156500           1370   98105   
867365  220750.0   450.0   05/19/1993     199000           1430   98102   
867366  546878.0   130.0   03/23/2005     309900           1940   98042   
867367  251700.0   900.0   07/05/2005     272195           1920   98042   
867368  881990.0   515.0   09/25/1985      63500            870   98105   
867369  222505.0  9207.0   10/09/1995     246500           2570   98005   
867370  222505.0  9207.0   10/09/1995     246500           1070   98005   
867371  541860.0    20.0   07/23/2002     450000           2890   98115   
867372  172606.0  9181.0   07/26/1983      87950           2320   98077   
867373  293500.0   100.0   01/23/1987      66850           1190   98001   
867374   80900.0  1440.0   10/19/1988     147000           1730   98109   
867375  247338.0   310.0   07/27/1987      83000           1870   98058   
867376   11410.0   629.0   01/26/1998     171000            730   98028   
867377  568000.0   570.0   01/09/201

### 3. Add a column for PricePerSqFt



In [76]:
sales_data['PricePerSqFt'] = sales_data['SalePrice']/sales_data['SqFtTotLiving']
sales_data

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0       138860.0   110.0   08/21/2014     245000           1490   98002   
1       138860.0   110.0   06/12/1989     109300           1490   98002   
2       138860.0   110.0   01/16/2005      14684           1490   98002   
3       423943.0    50.0   07/15/1999      96000            960   98092   
4       423943.0    50.0   01/08/2001     127500            960   98092   
5       403700.0   715.0   07/03/2013     464500           1780   98008   
6       403700.0   715.0   02/21/2013     357000           1780   98008   
7       403700.0   715.0   10/13/1995     142000           1780   98008   
8       403700.0   715.0   02/22/2007     528000           1780   98008   
9        98400.0   380.0   03/23/1994     139950           2050   98058   
10       98400.0   380.0   03/29/2017     439500           2050   98058   
11      330395.0  1610.0   06/07/2012     335000           2210   98038   
12      330395.0  1610.0   08/26/2013     369900           2210   98038   
13      330395.0  1610.0   06/27/1994     192950           2210   98038   
14      330395.0  1610.0   08/22/2013     369900           2210   98038   
15      330395.0  1610.0   11/10/1993     107500           2210   98038   
16      675260.0   940.0   03/06/1998     209950           2040   98031   
17      675260.0   940.0   08/01/1995     191950           2040   98031   
18      675260.0   940.0   09/13/2004     289950           2040   98031   
19      795508.0   170.0   04/09/2001     174000            980   98058   
20      795508.0   170.0   10/31/1996     135000            980   98058   
21      795508.0   170.0   04/30/2004     210000            980   98058   
22      795508.0   170.0   12/21/2012     148000            980   98058   
23      638580.0   110.0   07/14/2015     190000           2130   98188   
24      638580.0   110.0   12/21/2015     338000           2130   98188   
25      252206.0  9011.0   02/21/1997     300000           2530   98051   
26      252206.0  9011.0   11/28/2000     390000           2530   98051   
27      321165.0    50.0   05/10/2004    2100000           1800   98001   
28      321165.0    50.0   10/24/2013     215000           1800   98001   
29      321165.0    50.0   09/20/2004     270000           1800   98001   
...          ...     ...          ...        ...            ...     ...   
867356  928170.0     5.0   12/11/1986      89950           1630   98116   
867357  783280.0    60.0   08/21/1986      67500           1200   98146   
867358  809260.0    80.0   08/20/2001     145000           1170   98042   
867359  344800.0  1500.0   07/12/1996     174000            680   98125   
867360  177421.0   370.0   06/24/1993     233000           2930   98077   
867361  278731.0   900.0   10/18/1982      78000           1590   98031   
867362   12605.0  9008.0   10/01/1993     110000            790   98072   
867363   42506.0  9158.0   03/07/1988      35000           1920   98053   
867364  955020.0  4120.0   03/24/1994     156500           1370   98105   
867365  220750.0   450.0   05/19/1993     199000           1430   98102   
867366  546878.0   130.0   03/23/2005     309900           1940   98042   
867367  251700.0   900.0   07/05/2005     272195           1920   98042   
867368  881990.0   515.0   09/25/1985      63500            870   98105   
867369  222505.0  9207.0   10/09/1995     246500           2570   98005   
867370  222505.0  9207.0   10/09/1995     246500           1070   98005   
867371  541860.0    20.0   07/23/2002     450000           2890   98115   
867372  172606.0  9181.0   07/26/1983      87950           2320   98077   
867373  293500.0   100.0   01/23/1987      66850           1190   98001   
867374   80900.0  1440.0   10/19/1988     147000           1730   98109   
867375  247338.0   310.0   07/27/1987      83000           1870   98058   
867376   11410.0   629.0   01/26/1998     171000            730   98028   
867377  568000.0   570.0   01/09/201

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [77]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 867386 entries, 0 to 867385
Data columns (total 7 columns):
Major            867386 non-null float64
Minor            867386 non-null float64
DocumentDate     867386 non-null object
SalePrice        867386 non-null int64
SqFtTotLiving    867386 non-null int64
ZipCode          867386 non-null object
PricePerSqFt     867386 non-null float64
dtypes: float64(3), int64(2), object(2)
memory usage: 46.3+ MB


In [88]:
sales_data['Year'] = sales_data['DocumentDate'].str[-4:]
sales_data

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0       138860.0   110.0   08/21/2014     245000           1490   98002   
1       138860.0   110.0   06/12/1989     109300           1490   98002   
2       138860.0   110.0   01/16/2005      14684           1490   98002   
3       423943.0    50.0   07/15/1999      96000            960   98092   
4       423943.0    50.0   01/08/2001     127500            960   98092   
5       403700.0   715.0   07/03/2013     464500           1780   98008   
6       403700.0   715.0   02/21/2013     357000           1780   98008   
7       403700.0   715.0   10/13/1995     142000           1780   98008   
8       403700.0   715.0   02/22/2007     528000           1780   98008   
9        98400.0   380.0   03/23/1994     139950           2050   98058   
10       98400.0   380.0   03/29/2017     439500           2050   98058   
11      330395.0  1610.0   06/07/2012     335000           2210   98038   
12      330395.0  1610.0   08/26/2013     369900           2210   98038   
13      330395.0  1610.0   06/27/1994     192950           2210   98038   
14      330395.0  1610.0   08/22/2013     369900           2210   98038   
15      330395.0  1610.0   11/10/1993     107500           2210   98038   
16      675260.0   940.0   03/06/1998     209950           2040   98031   
17      675260.0   940.0   08/01/1995     191950           2040   98031   
18      675260.0   940.0   09/13/2004     289950           2040   98031   
19      795508.0   170.0   04/09/2001     174000            980   98058   
20      795508.0   170.0   10/31/1996     135000            980   98058   
21      795508.0   170.0   04/30/2004     210000            980   98058   
22      795508.0   170.0   12/21/2012     148000            980   98058   
23      638580.0   110.0   07/14/2015     190000           2130   98188   
24      638580.0   110.0   12/21/2015     338000           2130   98188   
25      252206.0  9011.0   02/21/1997     300000           2530   98051   
26      252206.0  9011.0   11/28/2000     390000           2530   98051   
27      321165.0    50.0   05/10/2004    2100000           1800   98001   
28      321165.0    50.0   10/24/2013     215000           1800   98001   
29      321165.0    50.0   09/20/2004     270000           1800   98001   
...          ...     ...          ...        ...            ...     ...   
867356  928170.0     5.0   12/11/1986      89950           1630   98116   
867357  783280.0    60.0   08/21/1986      67500           1200   98146   
867358  809260.0    80.0   08/20/2001     145000           1170   98042   
867359  344800.0  1500.0   07/12/1996     174000            680   98125   
867360  177421.0   370.0   06/24/1993     233000           2930   98077   
867361  278731.0   900.0   10/18/1982      78000           1590   98031   
867362   12605.0  9008.0   10/01/1993     110000            790   98072   
867363   42506.0  9158.0   03/07/1988      35000           1920   98053   
867364  955020.0  4120.0   03/24/1994     156500           1370   98105   
867365  220750.0   450.0   05/19/1993     199000           1430   98102   
867366  546878.0   130.0   03/23/2005     309900           1940   98042   
867367  251700.0   900.0   07/05/2005     272195           1920   98042   
867368  881990.0   515.0   09/25/1985      63500            870   98105   
867369  222505.0  9207.0   10/09/1995     246500           2570   98005   
867370  222505.0  9207.0   10/09/1995     246500           1070   98005   
867371  541860.0    20.0   07/23/2002     450000           2890   98115   
867372  172606.0  9181.0   07/26/1983      87950           2320   98077   
867373  293500.0   100.0   01/23/1987      66850           1190   98001   
867374   80900.0  1440.0   10/19/1988     147000           1730   98109   
867375  247338.0   310.0   07/27/1987      83000           1870   98058   
867376   11410.0   629.0   01/26/1998     171000            730   98028   
867377  568000.0   570.0   01/09/201

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [106]:
seattle_zips = pd.read_csv('Zip_Codes.csv')
seattle_zips['ZIP'] = seattle_zips['ZIP'].astype(str)

In [108]:

seattle_zip_list = list(seattle_zips['ZIP'])
seattle_zip_list

['98001',
 '98002',
 '98003',
 '98004',
 '98005',
 '98006',
 '98007',
 '98008',
 '98010',
 '98011',
 '98012',
 '98014',
 '98019',
 '98020',
 '98021',
 '98022',
 '98022',
 '98023',
 '98024',
 '98026',
 '98027',
 '98028',
 '98029',
 '98030',
 '98031',
 '98032',
 '98033',
 '98034',
 '98036',
 '98037',
 '98038',
 '98039',
 '98040',
 '98042',
 '98043',
 '98045',
 '98047',
 '98047',
 '98051',
 '98052',
 '98053',
 '98055',
 '98056',
 '98057',
 '98058',
 '98059',
 '98065',
 '98068',
 '98070',
 '98072',
 '98072',
 '98074',
 '98075',
 '98077',
 '98077',
 '98087',
 '98092',
 '98092',
 '98101',
 '98102',
 '98103',
 '98104',
 '98105',
 '98106',
 '98107',
 '98108',
 '98109',
 '98110',
 '98112',
 '98115',
 '98116',
 '98117',
 '98118',
 '98119',
 '98121',
 '98122',
 '98125',
 '98126',
 '98133',
 '98134',
 '98136',
 '98144',
 '98146',
 '98148',
 '98154',
 '98155',
 '98158',
 '98164',
 '98166',
 '98168',
 '98177',
 '98178',
 '98188',
 '98195',
 '98198',
 '98199',
 '98201',
 '98203',
 '98204',
 '98205',


In [110]:
sales_data_seattle = sales_data[sales_data['ZipCode'].isin(seattle_zip_list)]
sales_data_seattle

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0       138860.0   110.0   08/21/2014     245000           1490   98002   
1       138860.0   110.0   06/12/1989     109300           1490   98002   
2       138860.0   110.0   01/16/2005      14684           1490   98002   
3       423943.0    50.0   07/15/1999      96000            960   98092   
4       423943.0    50.0   01/08/2001     127500            960   98092   
5       403700.0   715.0   07/03/2013     464500           1780   98008   
6       403700.0   715.0   02/21/2013     357000           1780   98008   
7       403700.0   715.0   10/13/1995     142000           1780   98008   
8       403700.0   715.0   02/22/2007     528000           1780   98008   
9        98400.0   380.0   03/23/1994     139950           2050   98058   
10       98400.0   380.0   03/29/2017     439500           2050   98058   
11      330395.0  1610.0   06/07/2012     335000           2210   98038   
12      330395.0  1610.0   08/26/2013     369900           2210   98038   
13      330395.0  1610.0   06/27/1994     192950           2210   98038   
14      330395.0  1610.0   08/22/2013     369900           2210   98038   
15      330395.0  1610.0   11/10/1993     107500           2210   98038   
16      675260.0   940.0   03/06/1998     209950           2040   98031   
17      675260.0   940.0   08/01/1995     191950           2040   98031   
18      675260.0   940.0   09/13/2004     289950           2040   98031   
19      795508.0   170.0   04/09/2001     174000            980   98058   
20      795508.0   170.0   10/31/1996     135000            980   98058   
21      795508.0   170.0   04/30/2004     210000            980   98058   
22      795508.0   170.0   12/21/2012     148000            980   98058   
23      638580.0   110.0   07/14/2015     190000           2130   98188   
24      638580.0   110.0   12/21/2015     338000           2130   98188   
25      252206.0  9011.0   02/21/1997     300000           2530   98051   
26      252206.0  9011.0   11/28/2000     390000           2530   98051   
27      321165.0    50.0   05/10/2004    2100000           1800   98001   
28      321165.0    50.0   10/24/2013     215000           1800   98001   
29      321165.0    50.0   09/20/2004     270000           1800   98001   
...          ...     ...          ...        ...            ...     ...   
867354  882290.0    90.0   10/11/1985      81750           2700   98125   
867355  149830.0  1035.0   08/10/1988      60000           1140   98144   
867356  928170.0     5.0   12/11/1986      89950           1630   98116   
867357  783280.0    60.0   08/21/1986      67500           1200   98146   
867358  809260.0    80.0   08/20/2001     145000           1170   98042   
867359  344800.0  1500.0   07/12/1996     174000            680   98125   
867360  177421.0   370.0   06/24/1993     233000           2930   98077   
867361  278731.0   900.0   10/18/1982      78000           1590   98031   
867362   12605.0  9008.0   10/01/1993     110000            790   98072   
867363   42506.0  9158.0   03/07/1988      35000           1920   98053   
867364  955020.0  4120.0   03/24/1994     156500           1370   98105   
867365  220750.0   450.0   05/19/1993     199000           1430   98102   
867366  546878.0   130.0   03/23/2005     309900           1940   98042   
867367  251700.0   900.0   07/05/2005     272195           1920   98042   
867368  881990.0   515.0   09/25/1985      63500            870   98105   
867369  222505.0  9207.0   10/09/1995     246500           2570   98005   
867370  222505.0  9207.0   10/09/1995     246500           1070   98005   
867372  172606.0  9181.0   07/26/1983      87950           2320   98077   
867373  293500.0   100.0   01/23/1987      66850           1190   98001   
867374   80900.0  1440.0   10/19/1988     147000           1730   98109   
867375  247338.0   310.0   07/27/1987      83000           1870   98058   
867376   11410.0   629.0   01/26/199

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [111]:
sales_data_seattle['PricePerSqFt'].loc[sales_data_seattle['Year'] == '2019'].mean()

402.88988958518263